dotnet/src/Extensions/PromptTemplate.Handlebars/Helpers/SystemHelpers/KernelSystemHelpers.cs

```diff
@@ -0,0 +1,169 @@
+    {
+        RegisterHandlebarsDotNetHelpers(handlebarsInstance, options);
+        RegisterSystemHelpers(handlebarsInstance, variables);
+    }
+
+    private static void RegisterHandlebarsDotNetHelpers(IHandlebars handlebarsInstance, HandlebarsPromptTemplateOptions helperOptions)
+    {
+        HandlebarsHelpers.Register(handlebarsInstance, optionsCallback: options =>
+        {
+            options.PrefixSeparator = helperOptions.PrefixSeparator;
+            options.Categories = helperOptions.Categories;
+            options.UseCategoryPrefix = helperOptions.UseCategoryPrefix;
+            options.CustomHelperPaths = helperOptions.CustomHelperPaths;
+        });
+    }
+
+    /// <summary>
+    /// Register all system helpers.
+    /// </summary>
+    /// <param name="handlebarsInstance">The <see cref="IHandlebars"/>-instance.</param>
+    /// <param name="variables">Dictionary of variables maintained by the Handlebars context.</param>
+    /// <exception cref="KernelException">Exception thrown when a message does not contain a defining role.</exception>
+    private static void RegisterSystemHelpers(
+        IHandlebars handlebarsInstance,
+        KernelArguments variables)
+    {
+        // TODO [@teresaqhoang]: Issue #3947 Isolate Handlebars Kernel System helpers in their own class
+        // Should also consider standardizing the naming conventions for these helpers, i.e., 'Message' instead of 'message'
+        handlebarsInstance.RegisterHelper("message", (writer, options, context, arguments) =>
+        {
+            var parameters = arguments[0] as IDictionary<string, object>;
+
+            // Verify that the message has a role
+            if (!parameters!.ContainsKey("role"))
+            {
+                throw new KernelException("Message must have a role.");
+            }
+
+            writer.Write($"<{parameters["role"]}~>", false);
+            options.Template(writer, context);
+            writer.Write($"</{parameters["role"]}~>", false);
+        });
+
+        handlebarsInstance.RegisterHelper("set", (writer, context, arguments) =>
+        {
+            var name = string.Empty;
+            object value = string.Empty;
+            if (arguments[0].GetType() == typeof(HashParameterDictionary))
+            {
+                // Get the parameters from the template arguments
+                var parameters = arguments[0] as IDictionary<string, object>;
+                name = (string)parameters!["name"];
+                value = parameters!["value"];
+            }
+            else
+            {
+                name = arguments[0].ToString();
+                value = arguments[1];
+            }
+
+            if (variables.TryGetValue(name, out var argVal))
+            {
+                variables[name] = value;
+            }
+            else
+            {
+                variables.Add(name, value);
+            }
```

Comment:
```suggestion
            variables[name] = value;
```
?

Reply:
If I remember correctly, this threw an error during testing if an entry with that key already existed in variables dict, (or maybe I was just using `.Add` before) -- will have to double test this

Reply:
> this threw an error during testing if an entry with that key already existed in variables dict

It shouldn't. It should overwrite an existing entry if it's already there or add a new entry if it doesn't.

Reply:
Awesome, changed